engine/execution/ingestion/engine.go

```diff
@@ -376,182 +323,242 @@ func (e *Engine) executeBlock(ctx context.Context, executableBlock *entity.Execu
+		// double check before applying the state delta
+		if bytes.Equal(eb.StartState, delta.ExecutableBlock.StartState) {
+			e.unit.Launch(func() {
+				e.applyStateDelta(delta)
+			})
+			return true
+		}
+
+		// if state delta is invalid, remove the delta and log error
+		e.log.Error().
+			Hex("block_start_state", eb.StartState).
+			Hex("delta_start_state", delta.ExecutableBlock.StartState).
+			Msg("can not apply the state delta, the start state does not match")
+
+		e.syncDeltas.Rem(eb.Block.ID())
+	}
+
+	// if don't have the delta, then check if everything is ready for executing
+	// the block
 	if eb.IsComplete() {
-		e.log.Debug().
-			Hex("block_id", logging.Entity(eb)).
-			Msg("block complete, starting execution")
 
 		if e.extensiveLogging {
 			e.logExecutableBlock(eb)
 		}
 
-		e.wg.Add(1)
-		go e.executeBlock(context.Background(), eb)
+		e.unit.Launch(func() {
+			e.executeBlock(e.unit.Ctx(), eb)
+		})
 		return true
 	}
 	return false
 }
 
+// OnCollection is a callback for handling the collections requested by the
+// collection requester.
 func (e *Engine) OnCollection(originID flow.Identifier, entity flow.Entity) {
-	err := e.handleCollection(originID, entity)
-	if err != nil {
-		e.log.Error().Err(err).Msg("could not handle collection")
-		return
-	}
-}
-
-func (e *Engine) handleCollection(originID flow.Identifier, entity flow.Entity) error {
-
 	// convert entity to strongly typed collection
 	collection, ok := entity.(*flow.Collection)
 	if !ok {
-		return fmt.Errorf("invalid entity type (%T)", entity)
+		e.log.Error().Msgf("invalid entity type (%T)", entity)
 	}
 
-	collID := collection.ID()
+	// no need to validate the origin ID, since the collection requester has
+	// checked the origin must be a collection node.
 
+	err := e.handleCollection(originID, collection)
+	if err != nil {
+		e.log.Error().Err(err).Msg("could not handle collection")
+	}
+}
+
+// a block can't be executed if its collection is missing.
+// since a collection can belong to multiple blocks, we need to
+// find all the blocks that are needing this collection, and then
+// check if any of these block becomes executable and execut it if
+// is.
+func (e *Engine) handleCollection(originID flow.Identifier, collection *flow.Collection) error {
+
+	// TODO: bail if have seen this collection before.
 	err := e.collections.Store(collection)
 	if err != nil {
 		return fmt.Errorf("cannot store collection: %w", err)
 	}
 
+	collID := collection.ID()
+
 	return e.mempool.BlockByCollection.Run(
 		func(backdata *stdmap.BlockByCollectionBackdata) error {
-			blockByCollectionId, exists := backdata.ByID(collID)
+			blockByCollectionID, exists := backdata.ByID(collID)
+
+			// if we don't find any block for this collection, then
+			// means we don't need this collection any more.
+			// or it was ejected from the mempool when it was full.
+			// either way, we will return
 			if !exists {
-				return fmt.Errorf("could not find block for collection")
+				return nil
```

Comment:
We should log this situation, as this shouldn't happen, probably `debug` level would suffice